server/eclipse-project/src/main/java/us/freeandfair/corla/auth/AbstractAuthentication.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Free & Fair Colorado RLA System
+ * 
+ * @title ColoradoRLA
+ * @created Aug 29, 2017
+ * @copyright 2017 Free & Fair
+ * @license GNU General Public License 3.0
+ * @author Joe Kiniry <kiniry@freeandfair.us>
+ * @description A system to assist in conducting statewide risk-limiting audits.
+ */
+
+package us.freeandfair.corla.auth;
+
+import javax.persistence.PersistenceException;
+
+import spark.Request;
+import spark.Response;
+
+import us.freeandfair.corla.Main;
+import us.freeandfair.corla.model.Administrator;
+import us.freeandfair.corla.model.Administrator.AdministratorType;
+import us.freeandfair.corla.model.County;
+import us.freeandfair.corla.persistence.Persistence;
+import us.freeandfair.corla.query.AdministratorQueries;
+import us.freeandfair.corla.query.CountyQueries;
+
+/**
+ * An abstract base class that enforces the two-stage state machine for two-factor
+ * authentication.
+ * 
+ * @author Joseph R. Kiniry
+ */
+@SuppressWarnings({"PMD.AtLeastOneConstructor", "PMD.CyclomaticComplexity",
+    "PMD.ModifiedCyclomaticComplexity", "PMD.StdCyclomaticComplexity"})
+public abstract class AbstractAuthentication implements AuthenticationInterface {  
+  /**
+   * Authenticate the administrator `the_username` with credentials
+   * `the_password` (for traditional authentication) or `the_second_factor`
+   * (for two-factor authentication).  This method should be called twice in
+   * succession, first for traditional authentication and second for two-factor
+   * authentication.
+   * 
+   * @trace authentication.authenticate_county_administrator
+   * @trace authentication.authenticate_state_administrator
+   * @return true iff authentication succeeds.
+   * @param the_request The request.
+   * @param the_admin_type the type of administrator to attempt to authenticate.
+   * @param the_username the username of the person to attempt to authenticate.
+   * @param the_password the password for `username`.
+   * @param the_second_factor the second factor for `username`.
+   */
+  @Override
+  public boolean authenticateAdministrator(final Request the_request,
+                                           final Response the_response,
+                                           final AdministratorType the_admin_type,
```

Comment:
We probably want to eliminate the AdministratorType from this parameter list at this point, since it looks to me as though it is in fact not used for anything.

Reply:
I kept it because I imagine some deployments may be using different authentication server and services for different roles.  E.g., public might use an HTTP mechanism, while administrators might use some kind of 2FA.  Indeed, we are not using it for this particular deployment.